File: backend-dotnet/Backend/Application/DTO/User/UpdateUserRequest.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace Backend.Application.DTO.User;

public record UpdateUserRequest(
    [property: JsonPropertyName("name")]
    string? Name,

    [property: JsonPropertyName("email")]
    [param: ValidEmailIfPresent]
    string? Email,

    [property: JsonPropertyName("password")]
    string? Password
);

public class ValidEmailIfPresentAttribute : ValidationAttribute
{
    protected override ValidationResult? IsValid(object? value, ValidationContext validationContext)
    {
        if (value is null || string.IsNullOrWhiteSpace(value.ToString()))
            return ValidationResult.Success;

        var email = value.ToString()!;
        var isValid = new EmailAddressAttribute().IsValid(email);

        return isValid
            ? ValidationResult.Success
            : new ValidationResult("Email is not valid");
    }
}

File: backend-spring/src/main/java/io/github/agussyahrilmubarok/backend/application/dto/user/UpdateUserRequest.java
package io.github.agussyahrilmubarok.backend.application.dto.user;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.validation.constraints.Email;

public record UpdateUserRequest(
        @JsonProperty("name") String name,

        @JsonProperty("email") @Email(message = "Email is not valid")
        String email,

        @JsonProperty("password") String password) {}
